isthmus/src/main/java/io/substrait/isthmus/OuterReferenceResolver.java

```diff
@@ -0,0 +1,88 @@
+package io.substrait.isthmus;
+
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.CorrelationId;
+import org.apache.calcite.rel.logical.LogicalFilter;
+import org.apache.calcite.rex.*;
+
+/** Resolve correlated variable and get Depth map for RexFieldAccess */
```

Comment:
My gut is that you did a bunch of research to figure out all of this. Maybe put a more in-depth comment here (and elsewhere as appropriate)? The steps out thing especially seemed like something that was initially confusing. I'm guessing if someone comes in here later, they're going to need help to fix a bug.

Reply:
I put explanation how the code logic works in `OuterReferenceResolver.md` file, with an example and how the steps work. 

Previously, I tried to put in-depth comment. But the java formatter changed the text-based tree graph, and make it hard to read.